### PR TITLE
CORE-170: rspace: implement getCheckpoint

### PR DESCRIPTION
**Pull request by henrytill**
## Overview
This PR implements `getCheckpoint`, a new part of the rspace API.

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/CORE-170

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
N/A

### TODO
- [x] Write tests
- [x] Add scaladoc
- [ ] ~Add comment about assumptions about transactions, locking, and writing to `_trieUpdates`~
- [x] Add TODO/comment about pruning `_trieUpdates`.

### CODE REVIEW

**Comment by lukasz-golebiewski** [comment]
Are you planning to write any tests?

File: rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala [hunk 1]
```diff
@@ -181,6 +181,8 @@ class InMemoryStore[C, P, A, K] private (
         val wks  = _waitingContinuations.getOrElse(hash, Seq.empty[WaitingContinuation[P, K]])
         (cs, Row(data, wks))
     }.toMap
+
+  def getCheckpoint(): Blake2b256Hash = throw new Exception("unimplemented")
```
Comment (lukasz-golebiewski) [inline rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala]: `???` is shorthand for that

**Comment by henrytill** [comment]
> Are you planning to write any tests?

Yes, mentioned [here](https://github.com/rchain/rchain/pull/860#issue-191616931). 😉

File: rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala (hunk 1 above)
Comment (henrytill) [reply rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala]: I wanted to be a little more explicit about the fact that I didn't just accidentally forget to implement this.  But yes, you are right.

**Comment by henrytill** [comment]
This is ready for further review, if anyone else is so inclined.

@dzajkowski: I had to make some minor changes to `IStore` and `InMemoryStore` to remove some vestigial complexity and alleviate testing headaches.  Namely, I dropped the abstract `H` type from `IStore` and just used `Blake2b256Hash` (we aren't going to be doing using different hashes across implementations, so may as well simplify).  Just as you are doing in your work, this necessitated using `Blake2b256Hash` as the key type in the existing `InMemoryStore`.  No big deal or anything - just wanted to give you a heads up.

File: rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala]: I wasn't aware of that, sorry for being a smartass

File: rspace/src/main/scala/coop/rchain/rspace/package.scala [hunk 2]
```diff
@@ -331,4 +332,15 @@ package object rspace {
...
+  /** Gets a checkpoint.  Yes, this is a handwave.
+    *
+    * @param store A store which satisfies the [[IStore]] interface.
+    * @tparam C A type representing a channel
+    * @tparam P A type representing a pattern
+    * @tparam A A type representing a piece of data
+    * @tparam K A type representing a continuation
+    * @return A BLAKE2b256 representing the checkpoint
+    */
+  def getCheckpoint[C, P, A, K](store: IStore[C, P, A, K]): Blake2b256Hash = store.getCheckpoint()
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/package.scala]: Wait. What's the reason for existence of this method?

File: rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala (hunk 1 above)
Comment (henrytill) [reply rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala]: No need to apologize - I should be the one apologizing for my "But yes", which now reads a bit heavy-handed.

I would much rather you give comments like this than withhold them.  So thank you.

File: rspace/src/main/scala/coop/rchain/rspace/package.scala (hunk 2 above)
Comment (henrytill) [reply rspace/src/main/scala/coop/rchain/rspace/package.scala]: I know, seems a bit redundant?  

From a public API perspective, so far we haven't had clients operate directly on the store object, other than to close it.   I wanted the public API functions for history and rollback to follow this convention until such time as we choose to refactor.

**Pull request merged**